dm/worker/log_test.go

```diff
@@ -0,0 +1,308 @@
+
+	db, _ := testSetUpDB(c)
+	defer db.Close()
+	p, err = LoadHandledPointer(db)
+	c.Assert(err, IsNil)
+	c.Assert(p.Location, Equals, int64(0))
+
+	p = Pointer{
+		Location: 1,
+	}
+
+	bs, _ := p.MarshalBinary()
+	c.Assert(db.Put(HandledPointerKey, bs, nil), IsNil)
+	p, err = LoadHandledPointer(db)
+	c.Assert(err, IsNil)
+	c.Assert(p.Location, Equals, int64(1))
+
+	c.Assert(db.Put(HandledPointerKey, []byte("xx"), nil), IsNil)
+	_, err = LoadHandledPointer(db)
+	c.Assert(err, ErrorMatches, ".*not valid length data as.*")
+}
+
+func (t *testLog) TestTaskLogKey(c *C) {
+	var id int64 = 1
+	idc, err := DecodeTaskLogKey(EncodeTaskLogKey(id))
+	c.Assert(err, IsNil)
+	c.Assert(idc, DeepEquals, id)
+
+	_, err = DecodeTaskLogKey([]byte("xx"))
+	c.Assert(err, ErrorMatches, ".*not valid length data as.*")
+}
+
+func (t *testLog) TestTaskLog(c *C) {
+	logger := new(Logger)
+
+	_, err := logger.Initial(nil)
+	c.Assert(err, Equals, ErrInValidHandler)
+
+	db, _ := testSetUpDB(c)
+	defer db.Close()
+	logs, err := logger.Initial(db)
+	c.Assert(logs, HasLen, 0)
+	c.Assert(err, IsNil)
+	c.Assert(logger.handledPointer.Location, Equals, int64(0))
+	c.Assert(logger.endPointer.Location, Equals, int64(1))
+
+	// try to get log from empty queue
+	l, err := logger.GetTaskLog(nil, 0)
+	c.Assert(err, Equals, ErrInValidHandler)
+
+	l, err = logger.GetTaskLog(db, 0)
+	c.Assert(err, IsNil)
+	c.Assert(l, IsNil)
+
+	// try to append log
+	c.Assert(logger.Append(nil, nil), Equals, ErrInValidHandler)
+
+	taskLog1 := &pb.TaskLog{
+		Id:   100,
+		Task: testTask1Meta,
+	}
+
+	c.Assert(logger.Append(db, taskLog1), IsNil)
+	c.Assert(taskLog1.Id, Equals, int64(1))
+	c.Assert(taskLog1.Ts > 0, IsTrue)
+	c.Assert(taskLog1.Task.Name, Equals, "task1")
+
+	taskLog2 := &pb.TaskLog{
+		Id:   200,
+		Task: testTask1Meta,
+	}
+
+	c.Assert(logger.Append(db, taskLog2), IsNil)
+	c.Assert(taskLog2.Id, Equals, int64(2))
+	c.Assert(taskLog2.Ts > 0, IsTrue)
+	c.Assert(taskLog2.Task.Name, Equals, "task1")
+
+	// try to get log
+	l, err = logger.GetTaskLog(db, 0)
+	c.Assert(err, IsNil)
+	c.Assert(l, IsNil)
+
+	l, err = logger.GetTaskLog(db, 1)
+	c.Assert(err, IsNil)
+	c.Assert(l, DeepEquals, taskLog1)
+
+	l, err = logger.GetTaskLog(db, 2)
+	c.Assert(err, IsNil)
+	c.Assert(l, DeepEquals, taskLog2)
+
+	logs, err = logger.Initial(db)
+	c.Assert(err, IsNil)
+	c.Assert(logs, DeepEquals, []*pb.TaskLog{taskLog1, taskLog2})
+	c.Assert(logger.handledPointer.Location, Equals, int64(0))
+	c.Assert(logger.endPointer.Location, Equals, int64(3))
+
+	// try to forward
+	c.Assert(logger.ForwardTo(db, 1), IsNil)
+	c.Assert(logger.handledPointer.Location, Equals, int64(1))
+	hp, err := LoadHandledPointer(db)
+	c.Assert(err, IsNil)
+	c.Assert(hp.Location, Equals, int64(1))
+
+	c.Assert(logger.ForwardTo(db, 2), IsNil)
+	c.Assert(logger.handledPointer.Location, Equals, int64(2))
+	hp, err = LoadHandledPointer(db)
+	c.Assert(err, IsNil)
+	c.Assert(hp.Location, Equals, int64(2))
+
+	c.Assert(logger.ForwardTo(db, 0), IsNil)
```

Comment:
do we need to support `ForwardTo` to _backward_?

Reply:
Depending on the caller